File: src/ep/ops/controlflow.cc
// Copyright (c) 2026. Licensed under the MIT License.
//
// Control-flow op handlers: If, Scan, Loop. Unlike ordinary ops these carry their computation as a
// nested subgraph (GraphProto) ATTRIBUTE — a body ORT surfaces to a plugin EP via Node_GetSubgraphs
// (ep.cc::BuildSubgraphs captures each body as a NodeDesc::subgraphs entry). The MLX-native EP owns
// the control-flow node WHOLE (its body is declined for independent offload in ep.cc::GetCapability)
// and realizes the control flow by translating the body inline through TranslationContext::RunSubgraph:
//
//   * If   — read the runtime `cond` host-side each forward and translate the taken branch only. Both
//            branches must be translatable (either can be taken across forwards); the graph is rebuilt
//            per forward (eager path), so per-forward branch selection is exact. A runtime `cond` needs
//            no mlx_where because we pick the branch at graph-build time.
//   * Scan — STATIC trip count (scan axis length known from the input shape). Unroll the body over the
//            axis, carrying state and stacking scan outputs. Forward direction, axis 0 only (MVP).
//   * Loop — CONSTANT trip count M with a cond that is a pass-through of the loop cond input (the
//            canonical `for i in range(M)` idiom, provably a fixed M-iteration loop). Unroll M times.
//            Carried-state only (no per-iteration scan outputs) in this MVP. Dynamic / cond-dependent
//            trip counts are NOT statically unrollable and are left to ORT CPU (unclaimed).
//
// Everything claimed here is translated correctly; anything outside these static/foldable forms is
// left unclaimed and runs on ORT's CPU control-flow kernels (with the body ops still offloaded to MLX
// via the ordinary flat path — see ep.cc::GetCapability).

#include <cstdint>
#include <string>
#include <vector>

#include "mlx_engine.h"
#include "op_claim.h"
#include "op_registry.h"

namespace ort_mlx {

namespace {

// Every node in a control-flow body must be MLX-translatable (recursively, so a nested If/Scan/Loop
// is checked through its own claim predicate). Consulted at claim time so a claimed control-flow node
// can never contain an untranslatable op.
bool BodyClaimable(Ort::ConstGraph body) {
  for (Ort::ConstNode node : body.GetNodes()) {
    if (!Claimable(node)) return false;
  }
  return true;
}

// Find a body subgraph by attribute name.
const SubgraphDesc* FindBody(const NodeDesc& n, const char* attr) {
  for (const SubgraphDesc& sg : n.subgraphs) {
    if (sg.attr_name == attr) return &sg;
  }
  return nullptr;
}

// Read a scalar bool from a foldable (initializer / ctx) node input.
bool ReadHostBool(TranslationContext& ctx, const TensorRef& ref) {
  HostBytes h = ctx.RawHost(ref);
  return h.data != nullptr && *reinterpret_cast<const uint8_t*>(h.data) != 0;
}

// Stack a list of same-shaped MLX arrays along a new leading axis (axis 0), like ONNX scan-output
// accumulation. Keeps + returns the result.
mlx_array StackAxis0(TranslationContext& ctx, const std::vector<mlx_array>& parts) {
  mlx_vector_array vec = mlx_vector_array_new();
  for (mlx_array a : parts) mlx_vector_array_append_value(vec, a);
  mlx_array r = mlx_array_new();
  int rc = mlx_stack_axis(&r, vec, 0, ctx.stream());
  mlx_vector_array_free(vec);
  if (rc != 0) throw MlxError("MLX Scan/Loop: mlx_stack_axis failed");
  return ctx.Keep(r);
}

// ---- If ---------------------------------------------------------------------------------------

void IfOp(TranslationContext& ctx, const NodeDesc& n) {
  const bool cond = ReadHostBool(ctx, n.inputs[0]);
  const SubgraphDesc* branch = FindBody(n, cond ? "then_branch" : "else_branch");
  if (branch == nullptr) throw MlxError("MLX If: missing branch subgraph");
  std::vector<mlx_array> outs = ctx.RunSubgraph(*branch, {});
  if (outs.size() != n.outputs.size()) throw MlxError("MLX If: branch output arity mismatch");
  for (size_t i = 0; i < n.outputs.size(); ++i) ctx.Bind(n.outputs[i], outs[i]);
}

bool IfClaim(Ort::ConstNode node) {
  const std::vector<Ort::ConstValueInfo> inputs = node.GetInputs();
  const std::vector<Ort::ConstValueInfo> outputs = node.GetOutputs();
  if (inputs.size() != 1 || outputs.empty()) return false;
  ONNXTensorElementDataType cond_type;
  if (!TensorInfo(inputs[0], cond_type) || cond_type != ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL) {
    return false;
  }
  std::vector<Ort::AttrNameSubgraph> subs = node.GetSubgraphs();
  if (subs.size() != 2) return false;
  bool have_then = false, have_else = false;
  for (Ort::AttrNameSubgraph& as : subs) {
    if (as.attr_name == "then_branch") have_then = true;
    else if (as.attr_name == "else_branch") have_else = true;
    else return false;
    Ort::ConstGraph body = as.sub_graph;
    if (!body.GetInputs().empty()) return false;                 // If branches take no formal inputs
    if (body.GetOutputs().size() != outputs.size()) return false;
    if (!BodyClaimable(body)) return false;
  }
  return have_then && have_else;
}

// ---- Scan -------------------------------------------------------------------------------------

// Reject non-default (non-forward / non-axis-0) direction/axis attributes — MVP supports forward
// iteration over axis 0 only.
bool AllZeroIntsAttr(Ort::ConstNode node, const char* name) {
  std::vector<int64_t> v;
  bool present = false;
  IntsAttribute(node, name, v, present);
  if (!present) return true;
  for (int64_t x : v) {
    if (x != 0) return false;
  }
  return true;
}

void ScanOp(TranslationContext& ctx, const NodeDesc& n) {
  const int64_t num_scan = n.ints.at("num_scan_inputs");
  const int num_state = static_cast<int>(n.inputs.size()) - static_cast<int>(num_scan);
  const SubgraphDesc* body = FindBody(n, "body");
  if (body == nullptr) throw MlxError("MLX Scan: missing body subgraph");

  std::vector<mlx_array> state;
  for (int i = 0; i < num_state; ++i) state.push_back(ctx.Resolve(n.inputs[i]));
  std::vector<mlx_array> scans;
  for (int i = 0; i < static_cast<int>(num_scan); ++i) {
    scans.push_back(ctx.Resolve(n.inputs[num_state + i]));
  }

  const std::vector<int> s0 = TranslationContext::ShapeOf(scans[0]);
  if (s0.empty()) throw MlxError("MLX Scan: scan input is a scalar");
  const int trip = s0[0];

  const int num_scan_out = static_cast<int>(body->output_names.size()) - num_state;
  if (num_scan_out < 0) throw MlxError("MLX Scan: body output arity");
  std::vector<std::vector<mlx_array>> collected(num_scan_out);

  for (int t = 0; t < trip; ++t) {
    std::vector<mlx_array> bin;
    bin.reserve(num_state + num_scan);
    for (int i = 0; i < num_state; ++i) bin.push_back(state[i]);
    for (int i = 0; i < static_cast<int>(num_scan); ++i) {
      const std::vector<int> shp = TranslationContext::ShapeOf(scans[i]);
      std::vector<int> start(shp.size(), 0), stop = shp;
      start[0] = t;
      stop[0] = t + 1;
      mlx_array sl = ctx.Slice(scans[i], start, stop);  // [1, ...]
      mlx_array sq = mlx_array_new();
      if (mlx_squeeze_axis(&sq, sl, 0, ctx.stream()) != 0) {
        throw MlxError("MLX Scan: mlx_squeeze_axis failed");
      }
      bin.push_back(ctx.Keep(sq));
    }
    std::vector<mlx_array> bout = ctx.RunSubgraph(*body, bin);
    for (int i = 0; i < num_state; ++i) state[i] = bout[i];
    for (int i = 0; i < num_scan_out; ++i) collected[i].push_back(bout[num_state + i]);
  }

  for (int i = 0; i < num_state; ++i) ctx.Bind(n.outputs[i], state[i]);
  for (int i = 0; i < num_scan_out; ++i) {
    ctx.Bind(n.outputs[num_state + i], StackAxis0(ctx, collected[i]));
  }
}

bool ScanClaim(Ort::ConstNode node) {
  const std::vector<Ort::ConstValueInfo> inputs = node.GetInputs();
  const std::vector<Ort::ConstValueInfo> outputs = node.GetOutputs();
  const int64_t num_scan = IntAttribute(node, "num_scan_inputs", -1);
  if (num_scan <= 0 || static_cast<int64_t>(inputs.size()) < num_scan) return false;
  const int64_t num_state = static_cast<int64_t>(inputs.size()) - num_scan;
  if (num_state < 0) return false;

  if (!AllZeroIntsAttr(node, "scan_input_directions")) return false;
  if (!AllZeroIntsAttr(node, "scan_output_directions")) return false;
  if (!AllZeroIntsAttr(node, "scan_input_axes")) return false;
  if (!AllZeroIntsAttr(node, "scan_output_axes")) return false;

  // Every scan input must have a statically-known, non-empty scan axis (axis 0) for a static unroll.
  for (int64_t i = num_state; i < static_cast<int64_t>(inputs.size()); ++i) {
    ONNXTensorElementDataType t;
    std::vector<int64_t> shp;
    if (!TensorInfo(inputs[i], t, &shp)) return false;
    if (shp.empty() || shp[0] < 1) return false;
  }

  std::vector<Ort::AttrNameSubgraph> subs = node.GetSubgraphs();
  if (subs.size() != 1 || subs[0].attr_name != "body") return false;
  Ort::ConstGraph body = subs[0].sub_graph;
  if (static_cast<int64_t>(body.GetInputs().size()) != num_state + num_scan) return false;
  if (static_cast<int64_t>(body.GetOutputs().size()) < num_state) return false;
  // The node exposes carried state + every scan output (= all body outputs beyond the state).
  if (static_cast<int64_t>(outputs.size()) != static_cast<int64_t>(body.GetOutputs().size())) {
    return false;
  }
  return BodyClaimable(body);
}

// ---- Loop -------------------------------------------------------------------------------------

// True iff the body's cond output (body output 0) is a pass-through of the body's cond input (body
// input 1): either a direct graph-output alias, or an Identity node copying it. With an initial cond
// of true this makes the loop run EXACTLY M iterations (no data-dependent early exit), so a constant M
// is a static trip count we can unroll.
bool LoopCondIsPassthrough(Ort::ConstGraph body) {
  const std::vector<Ort::ConstValueInfo> bin = body.GetInputs();
  const std::vector<Ort::ConstValueInfo> bout = body.GetOutputs();
  if (bin.size() < 2 || bout.empty()) return false;
  const std::string cond_in = bin[1].GetName();
  const std::string cond_out = bout[0].GetName();
  if (cond_in.empty() || cond_out.empty()) return false;
  if (cond_in == cond_out) return true;  // graph output directly aliases the cond input
  for (Ort::ConstNode node : body.GetNodes()) {
    if (node.GetOperatorType() != "Identity") continue;
    const std::vector<Ort::ConstValueInfo> ins = node.GetInputs();
    const std::vector<Ort::ConstValueInfo> outs = node.GetOutputs();
    if (ins.size() == 1 && outs.size() == 1 && ins[0].GetName() == cond_in &&
        outs[0].GetName() == cond_out) {
      return true;
    }
  }
  return false;
}

void LoopOp(TranslationContext& ctx, const NodeDesc& n) {
  const SubgraphDesc* body = FindBody(n, "body");
  if (body == nullptr) throw MlxError("MLX Loop: missing body subgraph");
  const int num_state = static_cast<int>(n.inputs.size()) - 2;  // inputs = [M, cond, state...]

  int64_t trip_count = 0;
  {
    HostBytes h = ctx.RawHost(n.inputs[0]);  // M (int64 scalar)
    if (h.data == nullptr) throw MlxError("MLX Loop: null trip count");
    trip_count = *reinterpret_cast<const int64_t*>(h.data);
  }
  const bool cond0 = ReadHostBool(ctx, n.inputs[1]);
  const int trip = cond0 ? static_cast<int>(trip_count) : 0;

  std::vector<mlx_array> state;
  for (int i = 0; i < num_state; ++i) state.push_back(ctx.Resolve(n.inputs[2 + i]));

  const bool cond_true = true;
  for (int t = 0; t < trip; ++t) {
    const int64_t iter_val = t;
    mlx_array iter = ctx.Keep(mlx_array_new_data(&iter_val, nullptr, 0, MLX_INT64));
    mlx_array condin = ctx.Keep(mlx_array_new_data(&cond_true, nullptr, 0, MLX_BOOL));
    std::vector<mlx_array> bin;
    bin.reserve(2 + num_state);
    bin.push_back(iter);
    bin.push_back(condin);
    for (int i = 0; i < num_state; ++i) bin.push_back(state[i]);
    std::vector<mlx_array> bout = ctx.RunSubgraph(*body, bin);
    // bout[0] = cond_out (pass-through, guaranteed true by the claim); bout[1..N] = carried state.
    for (int i = 0; i < num_state; ++i) state[i] = bout[1 + i];
  }

  for (int i = 0; i < num_state; ++i) ctx.Bind(n.outputs[i], state[i]);
}

bool LoopClaim(Ort::ConstNode node) {
  const std::vector<Ort::ConstValueInfo> inputs = node.GetInputs();
  const std::vector<Ort::ConstValueInfo> outputs = node.GetOutputs();
  if (inputs.size() < 2) return false;  // require explicit M and cond (MVP)
  const int num_state = static_cast<int>(inputs.size()) - 2;
  // M (int64 scalar) and cond (bool) must be present so we can read them host-side.
  ONNXTensorElementDataType mt, ct;
  if (!TensorInfo(inputs[0], mt) || mt != ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64) return false;
  if (!TensorInfo(inputs[1], ct) || ct != ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL) return false;

  std::vector<Ort::AttrNameSubgraph> subs = node.GetSubgraphs();
  if (subs.size() != 1 || subs[0].attr_name != "body") return false;
  Ort::ConstGraph body = subs[0].sub_graph;
  // body inputs = [iter_num, cond_in, state...]; carried-state-only (no scan outputs) in this MVP,
  // so body outputs = [cond_out, state...] and node outputs = [state...].
  if (static_cast<int>(body.GetInputs().size()) != 2 + num_state) return false;
  if (static_cast<int>(body.GetOutputs().size()) != 1 + num_state) return false;
  if (static_cast<int>(outputs.size()) != num_state) return false;
  if (!LoopCondIsPassthrough(body)) return false;
  return BodyClaimable(body);
}

}  // namespace

void RegisterControlFlowOps(OpRegistry& registry) {
  // ai.onnx control flow. Version-insensitive registration: the claim predicates gate the concrete
  // static/foldable forms; any node outside them is left to ORT CPU.
  registry.Register({"", "If", kAnyOpset, kAnyOpset, &IfOp, &IfClaim});
  registry.Register({"", "Scan", kAnyOpset, kAnyOpset, &ScanOp, &ScanClaim});
  registry.Register({"", "Loop", kAnyOpset, kAnyOpset, &LoopOp, &LoopClaim});
}

}  // namespace ort_mlx

File: rust/src/ops/controlflow.rs
//! Control-flow op handlers: If, Scan, Loop. Faithful port of the C++ `ops/controlflow.cc`.
//!
//! Unlike ordinary ops these carry their computation as a nested subgraph (GraphProto) ATTRIBUTE — a
//! body ORT surfaces to a plugin EP via `Node_GetSubgraphs` (`ep::build_subgraphs` captures each body
//! as a `NodeDesc::subgraphs` entry). The MLX EP owns the control-flow node WHOLE (its body is
//! declined for independent offload in `ep::get_capability`) and realizes the control flow by
//! translating the body inline through `TranslationContext::run_subgraph`:
//!
//!   * If   — read the runtime `cond` host-side each forward and translate the taken branch only.
//!   * Scan — STATIC trip count (scan axis length known from the input shape). Unroll the body over
//!            axis 0, carrying state and stacking scan outputs. Forward, axis 0 only (MVP).
//!   * Loop — CONSTANT trip count M with a cond that is a pass-through of the loop cond input (the
//!            canonical `for i in range(M)` idiom). Unroll M times; carried-state-only (MVP).
//!
//! Anything outside these static/foldable forms is left unclaimed and runs on ORT's CPU control-flow
//! kernels (with the body ops still offloaded to MLX via the ordinary flat path).

use crate::engine::{MlxError, NodeDesc, SubgraphDesc, TensorRef, TranslationContext};
use crate::registry::{
    ClaimPredicate, ClaimResult, GraphView, NodeView, OpHandler, OpRegistration, OpRegistry,
    K_ANY_OPSET,
};
use crate::sys::mlx;
use crate::sys::ort;
use crate::{deny, require};

// ---- shared helpers -----------------------------------------------------------------------------

/// Find a body subgraph by attribute name.
fn find_body<'a>(n: &'a NodeDesc, attr: &str) -> Option<&'a SubgraphDesc> {
    n.subgraphs.iter().find(|sg| sg.attr_name == attr)
}

/// Read a scalar bool from a foldable (initializer / ctx) node input.
fn read_host_bool(ctx: &TranslationContext, r: &TensorRef) -> Result<bool, MlxError> {
    let h = ctx.raw_host(r)?;
    if h.data.is_null() {
        return Ok(false);
    }
    Ok(unsafe { *(h.data as *const u8) } != 0)
}

/// Every node in a control-flow body must be MLX-translatable (recursively via the registry claim).
fn body_claimable(body: &GraphView) -> bool {
    body.all_nodes_claimable()
}

fn is_bool(node: &NodeView, i: usize) -> bool {
    matches!(node.input_info(i), Some(info)
        if info.dtype == ort::ONNXTensorElementDataType_ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL)
}

fn is_int64(node: &NodeView, i: usize) -> bool {
    matches!(node.input_info(i), Some(info)
        if info.dtype == ort::ONNXTensorElementDataType_ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64)
}

/// Reject non-default (non-forward / non-axis-0) direction/axis attributes.
fn all_zero_ints_attr(node: &NodeView, name: &str) -> bool {
    let (present, v) = node.ints_attr(name);
    if !present {
        return true;
    }
    v.iter().all(|&x| x == 0)
}

// ---- If -----------------------------------------------------------------------------------------

fn if_op(ctx: &mut TranslationContext, n: &NodeDesc) -> Result<(), MlxError> {
    let cond = read_host_bool(ctx, &n.inputs[0])?;
    let attr = if cond { "then_branch" } else { "else_branch" };
    let branch = find_body(n, attr)
        .ok_or_else(|| "MLX If: missing branch subgraph".to_string())?
        .clone();
    let outs = ctx.run_subgraph(&branch, &[])?;
    if outs.len() != n.outputs.len() {
        return Err("MLX If: branch output arity mismatch".to_string());
    }
    for (i, o) in n.outputs.iter().enumerate() {
        ctx.bind(o, outs[i]);
    }
    Ok(())
}

fn if_claim(node: &NodeView) -> ClaimResult {
    require!(
        node.num_inputs() == 1 && node.num_outputs() > 0,
        "expects 1 condition input and at least 1 output, got {}in/{}out",
        node.num_inputs(),
        node.num_outputs()
    );
    require!(is_bool(node, 0), "condition input must have bool dtype");
    let subs = node.subgraphs();
    require!(
        subs.len() == 2,
        "expects then_branch and else_branch subgraphs"
    );
    let (mut have_then, mut have_else) = (false, false);
    for (name, body) in &subs {
        match name.as_str() {
            "then_branch" => have_then = true,
            "else_branch" => have_else = true,
            _ => deny!("unsupported subgraph attribute {:?}", name),
        }
        require!(
            body.input_names().is_empty(),
            "{} must have no formal inputs",
            name
        );
        require!(
            body.output_names().len() == node.num_outputs(),
            "{} has {} outputs but the If node has {}",
            name,
            body.output_names().len(),
            node.num_outputs()
        );
        require!(
            body_claimable(body),
            "{} contains an unclaimable operation",
            name
        );
    }
    require!(
        have_then && have_else,
        "requires both then_branch and else_branch"
    );
    Ok(())
}

// ---- Scan ---------------------------------------------------------------------------------------

fn scan_op(ctx: &mut TranslationContext, n: &NodeDesc) -> Result<(), MlxError> {
    let num_scan =
        *n.ints
            .get("num_scan_inputs")
            .ok_or_else(|| "MLX Scan: missing num_scan_inputs".to_string())? as usize;
    let num_state = n.inputs.len() - num_scan;
    let body = find_body(n, "body")
        .ok_or_else(|| "MLX Scan: missing body subgraph".to_string())?
        .clone();

    let mut state: Vec<mlx::mlx_array> = Vec::with_capacity(num_state);
    for i in 0..num_state {
        state.push(ctx.resolve(&n.inputs[i])?);
    }
    let mut scans: Vec<mlx::mlx_array> = Vec::with_capacity(num_scan);
    for i in 0..num_scan {
        scans.push(ctx.resolve(&n.inputs[num_state + i])?);
    }

    let s0 = ctx.shape_of(scans[0]);
    if s0.is_empty() {
        return Err("MLX Scan: scan input is a scalar".to_string());
    }
    let trip = s0[0];

    let num_scan_out = body.output_names.len() as i64 - num_state as i64;
    if num_scan_out < 0 {
        return Err("MLX Scan: body output arity".to_string());
    }
    let num_scan_out = num_scan_out as usize;
    let mut collected: Vec<Vec<mlx::mlx_array>> = vec![Vec::new(); num_scan_out];

    for t in 0..trip {
        let mut bin: Vec<mlx::mlx_array> = Vec::with_capacity(num_state + num_scan);
        for i in 0..num_state {
            bin.push(state[i]);
        }
        for i in 0..num_scan {
            let shp = ctx.shape_of(scans[i]);
            let mut start = vec![0i32; shp.len()];
            let mut stop = shp.clone();
            start[0] = t;
            stop[0] = t + 1;
            let sl = ctx.slice(scans[i], &start, &stop)?;
            let sq = ctx.squeeze(sl, 0)?;
            bin.push(sq);
        }
        let bout = ctx.run_subgraph(&body, &bin)?;
        for i in 0..num_state {
            state[i] = bout[i];
        }
        for i in 0..num_scan_out {
            collected[i].push(bout[num_state + i]);
        }
    }

    for i in 0..num_state {
        ctx.bind(&n.outputs[i], state[i]);
    }
    for i in 0..num_scan_out {
        let stacked = ctx.stack(&collected[i], 0)?;
        ctx.bind(&n.outputs[num_state + i], stacked);
    }
    Ok(())
}

fn scan_claim(node: &NodeView) -> ClaimResult {
    let ninputs = node.num_inputs();
    let noutputs = node.num_outputs();
    let num_scan = node.int_attr("num_scan_inputs", -1);
    require!(
        num_scan > 0 && (ninputs as i64) >= num_scan,
        "num_scan_inputs must be positive and no greater than the {} inputs, got {}",
        ninputs,
        num_scan
    );
    let num_state = ninputs as i64 - num_scan;
    require!(num_state >= 0, "num_scan_inputs exceeds input count");

    require!(
        all_zero_ints_attr(node, "scan_input_directions"),
        "only forward scan_input_directions are supported"
    );
    require!(
        all_zero_ints_attr(node, "scan_output_directions"),
        "only forward scan_output_directions are supported"
    );
    require!(
        all_zero_ints_attr(node, "scan_input_axes"),
        "only scan_input_axes=0 is supported"
    );
    require!(
        all_zero_ints_attr(node, "scan_output_axes"),
        "only scan_output_axes=0 is supported"
    );

    for i in num_state..ninputs as i64 {
        let info = match node.input_info(i as usize) {
            Some(info) => info,
            None => deny!("scan input {} lacks tensor type/shape info", i),
        };
        require!(
            !info.shape.is_empty() && info.shape[0] >= 1,
            "scan input {} must have a statically known non-empty axis 0, got shape {:?}",
            i,
            info.shape
        );
    }

    let subs = node.subgraphs();
    require!(
        subs.len() == 1 && subs[0].0 == "body",
        "requires exactly one body subgraph"
    );
    let body = &subs[0].1;
    require!(
        body.input_names().len() as i64 == num_state + num_scan,
        "body has {} inputs, expected {} carried-state plus scan inputs",
        body.input_names().len(),
        num_state + num_scan
    );
    require!(
        (body.output_names().len() as i64) >= num_state,
        "body has {} outputs but requires at least {} carried-state outputs",
        body.output_names().len(),
        num_state
    );
    require!(
        noutputs == body.output_names().len(),
        "Scan has {} outputs but body has {}",
        noutputs,
        body.output_names().len()
    );
    require!(
        body_claimable(body),
        "body contains an unclaimable operation"
    );
    Ok(())
}

// ---- Loop ---------------------------------------------------------------------------------------

/// True iff the body's cond output (body output 0) is a pass-through of the body's cond input (body
/// input 1): either a direct graph-output alias, or an Identity node copying it.
fn loop_cond_is_passthrough(body: &GraphView) -> bool {
    let bin = body.input_names();
    let bout = body.output_names();
    if bin.len() < 2 || bout.is_empty() {
        return false;
    }
    let cond_in = &bin[1];
    let cond_out = &bout[0];
    if cond_in.is_empty() || cond_out.is_empty() {
        return false;
    }
    if cond_in == cond_out {
        return true;
    }
    for node in body.nodes() {
        if node.op_type() != "Identity" {
            continue;
        }
        let ins = node.input_names();
        let outs = node.output_names();
        if ins.len() == 1 && outs.len() == 1 && &ins[0] == cond_in && &outs[0] == cond_out {
            return true;
        }
    }
    false
}

fn loop_op(ctx: &mut TranslationContext, n: &NodeDesc) -> Result<(), MlxError> {
    let body = find_body(n, "body")
        .ok_or_else(|| "MLX Loop: missing body subgraph".to_string())?
        .clone();
    let num_state = n.inputs.len() - 2; // inputs = [M, cond, state...]

    let trip_count: i64 = {
        let h = ctx.raw_host(&n.inputs[0])?;
        if h.data.is_null() {
            return Err("MLX Loop: null trip count".to_string());
        }
        unsafe { *(h.data as *const i64) }
    };
    let cond0 = read_host_bool(ctx, &n.inputs[1])?;
    let trip = if cond0 { trip_count } else { 0 };

    let mut state: Vec<mlx::mlx_array> = Vec::with_capacity(num_state);
    for i in 0..num_state {
        state.push(ctx.resolve(&n.inputs[2 + i])?);
    }

    for t in 0..trip {
        let iter = ctx.scalar_i64(t);
        let condin = ctx.scalar_bool(true);
        let mut bin: Vec<mlx::mlx_array> = Vec::with_capacity(2 + num_state);
        bin.push(iter);
        bin.push(condin);
        for i in 0..num_state {
            bin.push(state[i]);
        }
        let bout = ctx.run_subgraph(&body, &bin)?;
        // bout[0] = cond_out (pass-through, guaranteed true by claim); bout[1..] = carried state.
        for i in 0..num_state {
            state[i] = bout[1 + i];
        }
    }

    for i in 0..num_state {
        ctx.bind(&n.outputs[i], state[i]);
    }
    Ok(())
}

fn loop_claim(node: &NodeView) -> ClaimResult {
    const UNSUPPORTED: &str = "Loop: only static carried-state loops (constant trip-count + passthrough cond, no scan outputs) are unrolled; scan outputs / dynamic control stay on CPU";
    let ninputs = node.num_inputs();
    require!(ninputs >= 2, "{}", UNSUPPORTED);
    let num_state = ninputs - 2;
    require!(is_int64(node, 0) && is_bool(node, 1), "{}", UNSUPPORTED);

    let subs = node.subgraphs();
    require!(subs.len() == 1 && subs[0].0 == "body", "{}", UNSUPPORTED);
    let body = &subs[0].1;
    require!(body.input_names().len() == 2 + num_state, "{}", UNSUPPORTED);
    require!(
        body.output_names().len() == 1 + num_state,
        "{}",
        UNSUPPORTED
    );
    require!(node.num_outputs() == num_state, "{}", UNSUPPORTED);
    require!(loop_cond_is_passthrough(body), "{}", UNSUPPORTED);
    require!(body_claimable(body), "{}", UNSUPPORTED);
    Ok(())
}

// ---- registration -------------------------------------------------------------------------------

fn reg(
    registry: &mut OpRegistry,
    op_type: &'static str,
    handler: OpHandler,
    claim: ClaimPredicate,
) {
    registry.register(OpRegistration {
        domain: "",
        op_type,
        min_opset: K_ANY_OPSET,
        max_opset: K_ANY_OPSET,
        handler,
        claim,
    });
}

pub fn register(registry: &mut OpRegistry) {
    reg(registry, "If", if_op, if_claim);
    reg(registry, "Scan", scan_op, scan_claim);
    reg(registry, "Loop", loop_op, loop_claim);
}
